### ISSUE TITLE
Function Key Management API is not case-sensitive , causing 404 as well as previous key overridden 

### ISSUE THREAD

**Issue by paraspatidar** [Azure/azure-functions-host#2687]
When using the Functions App Management API to get and set Function Keys ( https://[FunctionAppName].azurewebsites.net/admin/functions/[FunctionName]/keys/[KeyName] ) the In- and Output is caseSensitive, but the keys are overwritten if a key with the same name (not caseSensitive) already exists.

For example, if we want to look at two Function Keys: TestKey & testKey

See if key exists:

> GET  https://[FunctionAppName].azurewebsites.net/admin/functions/[FunctionName]/keys/TestKey
--> Response: 404

Create key with captial 'T':

> POST  https://[FunctionAppName].azurewebsites.net/admin/functions/[FunctionName]/keys/TestKey
--> Response: {..... TestKey}

See if Key exists now:

> GET  https://[FunctionAppName].azurewebsites.net/admin/functions/[FunctionName]/keys/TestKey
--> Response: {..... TestKey}

See if Key with small 't' exists:

> GET  https://[FunctionAppName].azurewebsites.net/admin/functions/[FunctionName]/keys/testKey
--> Response: 404

Create Key with small 't':

> POST https://[FunctionAppName].azurewebsites.net/admin/functions/[FunctionName]/keys/testKey
--> Response: {..... testKey}

See if both keys exist:

> GET  https://[FunctionAppName].azurewebsites.net/admin/functions/[FunctionName]/keys/TestKey
--> Response: 404

> GET  https://[FunctionAppName].azurewebsites.net/admin/functions/[FunctionName]/keys/testKey
--> Response: {... testKey}

_**Point to Note : if we create keys from Azure portal , then angular UI already takes care of that and doesn't allow to create duplicate keys irrespective of case/Ordinal.**_

#### Investigative information
-- POST mykey1
{"name":"mykey1","value":"jZMEduaHRZDDxC7P90MaYNGNmKjdMuq8ANaTd/UH7UVVB/5ZHAKMkQ==","links":[{"rel":"self","href":"https://fnzipdeploye.azurewebsites.net/admin/functions/HttpTriggerCSharp1/keys/mykey1"}]}

-- GET mykey1
{"name":"mykey1","value":"jZMEduaHRZDDxC7P90MaYNGNmKjdMuq8ANaTd/UH7UVVB/5ZHAKMkQ==","links":[{"rel":"self","href":"https://fnzipdeploye.azurewebsites.net/admin/functions/HttpTriggerCSharp1/keys/mykey1"}]}

-- GET KEYS
{"keys":[{"name":"abc","value":"/CMCsDahlK0dHTEjX12O4jqXbdx7llK0H0MgXhK55sy6a6X3H85vuw=="},
{"name":"my2","value":"rUPEOrpfH8LKhLRKVAAtb58aT0VP4egaGkST09j6TFbsMCjfDc1/VQ=="},
{"name":"my1","value":"G3xrlTWQy3NRMRYi7nVze5CALCuqdfcq1wesHxrPJPaM90ysRPTfZw=="},
{"name":"mykey1","value":"jZMEduaHRZDDxC7P90MaYNGNmKjdMuq8ANaTd/UH7UVVB/5ZHAKMkQ=="}],
"links":[{"rel":"self","href":"https://fnzipdeploye.azurewebsites.net/admin/functions/HttpTriggerCSharp1/keys/"}]}

-- POST Mykey1
{"name":"Mykey1","value":"Ii0ReLaxsdp4mNw7ooV4ENSb71e0NCZP85EhNe2vDhyzVtTUJoVs1A==","links":[{"rel":"self","href":"https://fnzipdeploye.azurewebsites.net/admin/functions/HttpTriggerCSharp1/keys/Mykey1"}]}

-- GET Mykey1
{"name":"Mykey1","value":"Ii0ReLaxsdp4mNw7ooV4ENSb71e0NCZP85EhNe2vDhyzVtTUJoVs1A==","links":[{"rel":"self","href":"https://fnzipdeploye.azurewebsites.net/admin/functions/HttpTriggerCSharp1/keys/Mykey1"}]}

-- GET KEYS
{"keys":[{"name":"abc","value":"/CMCsDahlK0dHTEjX12O4jqXbdx7llK0H0MgXhK55sy6a6X3H85vuw=="},
{"name":"my2","value":"rUPEOrpfH8LKhLRKVAAtb58aT0VP4egaGkST09j6TFbsMCjfDc1/VQ=="},
{"name":"my1","value":"G3xrlTWQy3NRMRYi7nVze5CALCuqdfcq1wesHxrPJPaM90ysRPTfZw=="},
{"name":"Mykey1","value":"Ii0ReLaxsdp4mNw7ooV4ENSb71e0NCZP85EhNe2vDhyzVtTUJoVs1A=="}],"links":[{"rel":"self","href":"https://fnzipdeploye.azurewebsites.net/admin/functions/HttpTriggerCSharp1/keys"}]}

-- GET mykey1
returns 404.

#### Expected behavior

1) Either should say that Key already exist and doesn't let create key using POST.
2)OR ,  if we want to maintain both keys , then it should not remove previous key from collection.

#### Related information in source Code

`[HttpPost]
        [Route("admin/functions/{name}/keys/{keyName}")]
        public Task<IActionResult> Post(string name, string keyName) => AddOrUpdateSecretAsync(keyName, null, name, ScriptSecretsType.Function);`

**this internally calls  , which is actually creating this new Key with case changed , but at the same time seems to be deleting old key.**

```
private async Task<KeyOperationResult> AddOrUpdateSecretAsync(ScriptSecretsType secretsType, string keyScope, string secretName, string secret, Func<ScriptSecrets> secretsFactory)
        {
            OperationResult result = OperationResult.NotFound;
            secret = secret ?? GenerateSecret();
            await ModifyFunctionSecretsAsync(secretsType, keyScope, secrets =>
            {
                Key key = secrets.GetFunctionKey(secretName, keyScope);

                var createAndUpdateKey = new Action<OperationResult>((o) =>
                {
                    var newKey = CreateKey(secretName, secret);
                    secrets.AddKey(newKey, keyScope);
                    result = o;
                });


                if (key == null)
                {
                    createAndUpdateKey(OperationResult.Created);
                }
                else if (secrets.RemoveKey(key, keyScope))
                {
                    createAndUpdateKey(OperationResult.Updated);
                }

                return secrets;
            }, secretsFactory);

            return new KeyOperationResult(secret, result);
        } 
````

**Comment by FinVamp1** [comment]
@fabiocav can you take a look and see if this is on the backlog?

**Comment by fabiocav** [comment]
Looks like a valid bug. Adding to triaged for a sprint assignment.

**Comment by paulbatum** [comment]
I agree there is a correctness issue here but considering that the UX already protects against this, the impact feels very low. @paraspatidar do you have any further comments on the impact here? i.e. if any customers encounter this issue, we can share this information and they can work around the problem by making sure all their keys have unique names.

**Comment by paulbatum** [comment]
Is it a breaking change to fix it? If it is not, then I'd prefer to schedule it for after V2 GA.

**Comment by paulbatum** [comment]
(unless you see this is as a trivial issue to fix)

**Comment by fabiocav** [comment]
This is a pretty simple issue to fix. And the fact that this is a change in API behavior really makes me want to have this done before GA

**Comment by fabiocav** [comment]
This would technically qualify as a breaking change, as we're changing API behavior that (correctly or not) people may depend on. It's unlikely in this case, but I would still like to make this change while the V2 preview window is open to avoid the behavior change after GA, when the product should be stable.

**Comment by paulbatum** [comment]
No longer meets the bar to be tracked as a should-have for V2 GA.
